mkosi/kmod.py

```diff
@@ -268,51 +251,102 @@ def resolve_module_dependencies(
             elif key == "firmware":
+                glob = "" if value.endswith("*") else ".*"
+                found = list(Path("usr/lib/firmware").glob(f"{value}{glob}"))
+                found = [fw for fw in found if fw.resolve().exists()]
                 if (Path("usr/lib/firmware") / value).exists():
-                    firmware.add(Path("usr/lib/firmware") / value)
+                    found.append(Path("usr/lib/firmware") / value)
 
-                glob = "" if value.endswith("*") else ".*"
+                if not found:
+                    missing_firmware.add(value)
+                    continue
 
-                firmware.update(Path("usr/lib/firmware").glob(f"{value}{glob}"))
+                firmware.update(found)
 
             elif key == "name":
-                # The file names use dashes, but the module names use underscores. We track the names in
-                # terms of the file names, since the depends use dashes and therefore filenames as well.
                 name = normalize_module_name(value)
 
-                moddep[name] = depends
-                firmwaredep[name] = firmware
+                for fw in missing_firmware:
+                    logging.warning(f"{fw} is a fw dependency of {name} but is not installed, ignoring")
+
+                result["moddep"][name] = depends
+                result["firmwaredep"][name] = firmware
 
                 depends = set()
                 firmware = set()
+                missing_firmware = set()
+
+    return result
 
-    todo = [*builtin, *modules]
-    mods = set()
-    firmware = set()
+
+def resolve_module_dependencies(
+    context: Context,
+    kver: str,
+    modules: Iterable[str],
+) -> tuple[set[Path], set[Path]]:
+    """Resolve module and firmware dependencies for specified modules
+
+    Returns a 2-tuple of sets of paths:
+    - The first set contains paths (relative to the root directory) to the
+      specified modules and all their direct and indirect module dependencies,
+      excluding built-in modules.
+    - The second set contains paths (relative to the root directory) to all
+      firmware files required by the modules in the first set.
+    """
+    modulesd = Path("usr/lib/modules") / kver
+    if (p := context.root / modulesd / "modules.builtin").exists():
+        builtin = {module_path_to_name(Path(m)) for m in p.read_text().splitlines()}
+    else:
+        builtin = set()
+    with chdir(context.root):
+        allmodules = set(modulesd.rglob("*.ko*"))
+    nametofile = {module_path_to_name(m): m for m in allmodules}
+
+    log_step("Resolving dependencies for kernel modules and detecting needed firmware")
+
+    moddep: dict[str, set[tuple[str, str]]] = {}
+    firmwaredep: dict[str, set[Path]] = {}
+
+    modules = {normalize_module_name(m) for m in modules}  # ensure normalized names
+    available = builtin | nametofile.keys()
+    todo = set(modules) & available
```

Comment:
`modules` is a set already.
```suggestion
    todo = modules & available
```